File: ios/Classes/HelpScoutApiImpl.swift
//
//  HelpScoutApi.swift
//  help_scout_sdk
//
//  Created by Mikel Corcuera on 23.02.2024.
//

import Foundation
import Beacon
import Flutter

extension FlutterError: Error {}

class HelpScoutApiImpl : NSObject, HelpScoutApi  {
  init(callbackApi: HelpScoutCallbackApi) {
    self.callbackApi = callbackApi
  }
  
  private var callbackApi: HelpScoutCallbackApi
  private var formPrefills: [String: HelpScoutApiFormPrefill] = [:]
  private var userSessionAttributes: [String: String] = [:]

  
  func openBeacon(beaconSettings: HelpScoutApiBeaconSettings) throws {
    let settings = convertBeaconSettings(beaconSettings)
    HSBeacon.open(settings)
  }
  
  func dismissBeacon() throws {
    HSBeacon.dismiss()
  }
  
  func reset() throws {
    HSBeacon.reset()
  }
  
  func navigate(settings: HelpScoutApiBeaconSettings, route: HelpScoutApiBeaconRoute) throws {
    let settings = convertBeaconSettings(settings)
    let route = switch route {
    case .answers: BeaconRoute.answers
    case .home:
      BeaconRoute.home
    case .ask:
      BeaconRoute.ask
    case .askMessage:
      BeaconRoute.askMessage
    case .askChat:
      BeaconRoute.askChat
    case .previousMessages:
      BeaconRoute.previousMessages
    }
    HSBeacon.navigate(route, settings: settings)
  }
  
  func search(settings: HelpScoutApiBeaconSettings, query: String) throws {
    let settings = convertBeaconSettings(settings)
    HSBeacon.search(query, beaconSettings: settings)
    
  }
  
  func openArticle(settings: HelpScoutApiBeaconSettings, articleId: String) throws {
    let settings = convertBeaconSettings(settings)
    HSBeacon.openArticle(articleId, beaconSettings: settings)
  }
  
  func prefillForm(beaconId: String, prefillData: HelpScoutApiFormPrefill) throws {
    formPrefills[beaconId] = prefillData
  }
  
  func resetAllFormPrefills() throws {
    HSBeacon.reset()
    formPrefills = [:]
  }
  
  func identify(user: HelpScoutApiUser) throws {
    let beaconUser = HSBeaconUser()
    beaconUser.email = user.email
    beaconUser.name = user.name
    beaconUser.company = user.company
    beaconUser.jobTitle = user.jobTitle
    if let avatar = user.avatar {
      beaconUser.avatar = URL(string: avatar)
    }
    if let attributes = user.attributes {
      for (key, value) in attributes {
        if let key = key, let value = value {
          beaconUser.addAttribute(withKey: key, value: value)
        }
      }
    }
    
    HSBeacon.identify(beaconUser)
  }
  
  func setSessionAttributes(attributes: [String? : String?]) throws {
    for (key, value) in attributes {
      if let key = key, let value = value {
        userSessionAttributes[key] = value;
      }
    }
  }
  
  
  
  func logout() throws {
    HSBeacon.logout()
  }
  
  private func convertBeaconSettings(_ beaconSettings: HelpScoutApiBeaconSettings) -> HSBeaconSettings {
    let settings = HSBeaconSettings(beaconId: beaconSettings.beaconId)
    
    if let messagingEnabled = beaconSettings.messagingEnabled {
      settings.messagingEnabled = messagingEnabled
    }
    
    if let focusModeOverride = beaconSettings.focusModeOverride {
      settings.focusModeOverride = switch(focusModeOverride) {
      case .neutral:
        HSBeaconFocusMode.neutral
      case .selfService:
        HSBeaconFocusMode.selfService
      case .askFirst:
        HSBeaconFocusMode.askFirst
      }
    }
    
    if let enablePreviousMessages = beaconSettings.enablePreviousMessages {
      settings.enablePreviousMessages = enablePreviousMessages
    }
    
    if let docsEnabled = beaconSettings.docsEnabled {
      settings.docsEnabled = docsEnabled
    }
    
    if let chatEnabled = beaconSettings.chatEnabled {
      settings.chatEnabled = chatEnabled
    }
    
    if let beaconTitle = beaconSettings.beaconTitle {
      settings.beaconTitle = beaconTitle
    }
    
    if let useNavigationBarAppearance = beaconSettings.useNavigationBarAppearance {
      settings.useNavigationBarAppearance = useNavigationBarAppearance
    }
    
    if let color = beaconSettings.color, let uiColor = UIColor(hex: color.hexCode) {
      settings.color = uiColor
    }
    
    if let tintColor = beaconSettings.tintColorOverride, let uiTintColor = UIColor(hex: tintColor.hexCode) {
      settings.tintColorOverride = uiTintColor
    }
    
    if let useLocalTranslationOverrides = beaconSettings.useLocalTranslationOverrides {
      settings.useLocalTranslationOverrides = useLocalTranslationOverrides
    }
    
    if let messagingSettings = beaconSettings.messagingSettings {
      if let contactFormCustomFieldsEnabled = messagingSettings.contactFormCustomFieldsEnabled {
        settings.messagingSettings.contactFormCustomFieldsEnabled = contactFormCustomFieldsEnabled
      }
      
      if let contactFormShowNameField = messagingSettings.contactFormShowNameField {
        settings.messagingSettings.contactFormShowNameField = contactFormShowNameField
      }
      
      if let contactFormShowSubjectField = messagingSettings.contactFormShowSubjectField {
        settings.messagingSettings.contactFormShowSubjectField = contactFormShowSubjectField
      }
      
      if let contactFormAllowAttachments = messagingSettings.contactFormAllowAttachments {
        settings.messagingSettings.contactFormAllowAttachments = contactFormAllowAttachments
      }
      
      if let showPrefilledCustomFields = messagingSettings.showPrefilledCustomFields {
        settings.messagingSettings.showPrefilledCustomFields = showPrefilledCustomFields
      }
    }
    
    settings.delegate = self
    return settings
  }
}

extension HelpScoutApiImpl : HSBeaconDelegate {
  func prefill(_ form: HSBeaconContactForm) {
    let beaconId = form.beaconID
    if let prefill = formPrefills[beaconId] {
      form.name = prefill.name
      if let email = prefill.email {
        form.email = email
      }
      form.subject = prefill.subject
      form.text = prefill.text
      
      if let customFields = prefill.customFields {
        for (id, value) in customFields {
          if let id = id, let value = value {
            form.addCustomFieldValue(value, forId: Int32(id))
          }
        }
      }
    }
  }
  
  func sessionAttributes() -> [String : String] {
    return userSessionAttributes
  }
  
  func onBeaconOpen(_ beaconSettings: HSBeaconSettings) {
    callbackApi.onBeaconOpen(beaconId: beaconSettings.beaconId, completion: {_ in })
  }
  
  func onBeaconInitialOpen(_ beaconSettings: HSBeaconSettings) {
      callbackApi.onBeaconInitialOpen(beaconId: beaconSettings.beaconId, completion: {_ in })
  }
  
  func onBeaconClose(_ beaconSettings: HSBeaconSettings) {
    callbackApi.onBeaconClose(beaconId: beaconSettings.beaconId, completion: {_ in })
  }
  
  func onBeaconInitialClose(_ beaconSettings: HSBeaconSettings) {
    callbackApi.onBeaconInitialClose(beaconId: beaconSettings.beaconId, completion: {_ in })
  }
}

extension UIColor {
  public convenience init?(hex: String) {
    let r, g, b, a: CGFloat
    print("Converting color \(hex)")
    if hex.hasPrefix("#") {
      let start = hex.index(hex.startIndex, offsetBy: 1)
      let hexColor = String(hex[start...])
      if hexColor.count != 6 && hexColor.count != 8 {
        return nil
      }
      var hexNumber: UInt64 = 0
      if Scanner(string: hexColor).scanHexInt64(&hexNumber) {
        if hexColor.count == 8 {
          a = CGFloat(hexNumber & 0x000000ff) / 255
          hexNumber = hexNumber >> 16
        } else {
          a = 1
        }
        
        r = CGFloat((hexNumber & 0xff0000) >> 16) / 255
        g = CGFloat((hexNumber & 0x00ff00) >> 8) / 255
        b = CGFloat((hexNumber & 0x0000ff) >> 0) / 255
        
        self.init(red: r, green: g, blue: b, alpha: a)
        return
      }
      
    }
    
    return nil
  }
}

File: android/src/main/kotlin/com/help_scout_sdk/HelpScoutApiImpl.kt
package com.help_scout_sdk

import HelpScoutApi
import HelpScoutApiBeaconRoute
import HelpScoutApiBeaconSettings
import HelpScoutApiFormPrefill
import HelpScoutApiUser
import HelpScoutCallbackApi
import android.app.Application
import com.helpscout.beacon.Beacon
import com.helpscout.beacon.internal.core.model.ContactFormConfigApi
import com.helpscout.beacon.model.BeaconConfigOverrides
import com.helpscout.beacon.model.FocusMode
import com.helpscout.beacon.ui.BeaconActivity
import android.content.Context
import android.content.Intent
import com.helpscout.beacon.model.BeaconScreens
import com.helpscout.beacon.model.PreFilledForm
import com.helpscout.beacon.ui.BeaconEventLifecycleHandler
import com.helpscout.beacon.ui.BeaconOnClosedListener
import com.helpscout.beacon.ui.BeaconOnOpenedListener
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

class HelpScoutApiImpl(private val applicationContext: Context, private val callbackApi: HelpScoutCallbackApi) :


    HelpScoutApi {
    private val eventLifecycleHandler = BeaconEventLifecycleHandler(
        object : BeaconOnOpenedListener {
            override fun onOpened() {
                if (currentBeaconId != null) {
                    callbackApi.onBeaconOpen(currentBeaconId!!) {}
                }
            }
        },
        object : BeaconOnClosedListener {
            override fun onClosed() {
                if (currentBeaconId != null) {
                    callbackApi.onBeaconClose(currentBeaconId!!) {}
                }
            }
        }
    )

    private var application: Application? = null
    private var currentBeaconId: String? =  null

    fun onAttachedToActivity(binding: ActivityPluginBinding) {
        application = binding.activity.application
        application?.registerActivityLifecycleCallbacks(eventLifecycleHandler)
    }

    fun onDetachedFromActivity() {
        application?.unregisterActivityLifecycleCallbacks(eventLifecycleHandler)
    }

    override fun openBeacon(beaconSettings: HelpScoutApiBeaconSettings) {
        configureBeacon(beaconSettings)
        BeaconActivity.open(applicationContext)
    }

    override fun dismissBeacon() {
    }

    override fun identify(beaconId: String, user: HelpScoutApiUser) {
        initBeacon(beaconId)
        Beacon.identify(user.email ?: "", user.name, user.company, user.jobTitle, user.avatar)
        if (user.attributes != null) {
            for ((key, value) in user.attributes) {
                if (key != null && value != null) {
                    Beacon.addAttributeWithKey(key, value)
                }
            }
        }
    }

    override fun setSessionAttributes(beaconId: String, attributes: Map<String?, String?>) {
        initBeacon(beaconId)
        Beacon.setSessionAttributes(attributes)
    }

    override fun prefillForm(beaconId: String, prefillData: HelpScoutApiFormPrefill) {
        val customFieldValues = mutableMapOf<Int, String>()

        if (prefillData.customFields != null) {
            for ((key, value) in prefillData.customFields) {
                if (key != null && value != null) {
                    customFieldValues[key.toInt()] = value
                }
            }
        }

        Beacon.addPreFilledForm(
            PreFilledForm(
                name = prefillData.name ?: "",
                subject = prefillData.subject ?: "",
                message = prefillData.text ?: "",
                email = prefillData.email ?: "",
                customFieldValues = customFieldValues
            )
        )
    }

    override fun resetFormPrefill(beaconId: String) {
        initBeacon(beaconId)
        Beacon.prefilledFormReset()
    }

    override fun navigate(settings: HelpScoutApiBeaconSettings, route: HelpScoutApiBeaconRoute) {
        configureBeacon(settings)
        BeaconActivity.open(
            applicationContext, when (route) {
                HelpScoutApiBeaconRoute.HOME -> BeaconScreens.DEFAULT
                HelpScoutApiBeaconRoute.ASK -> BeaconScreens.ASK
                HelpScoutApiBeaconRoute.ASK_MESSAGE -> BeaconScreens.CONTACT_FORM_SCREEN
                HelpScoutApiBeaconRoute.ASK_CHAT -> BeaconScreens.CHAT
                HelpScoutApiBeaconRoute.ANSWERS -> BeaconScreens.DEFAULT
                HelpScoutApiBeaconRoute.PREVIOUS_MESSAGES -> BeaconScreens.PREVIOUS_MESSAGES
            }, arrayListOf()
        )
    }

    override fun search(settings: HelpScoutApiBeaconSettings, query: String) {
        configureBeacon(settings)
        BeaconActivity.open(applicationContext, BeaconScreens.SEARCH_SCREEN, arrayListOf(query))
    }

    override fun openArticle(settings: HelpScoutApiBeaconSettings, articleId: String) {
        configureBeacon(settings)
        BeaconActivity.open(
            applicationContext,
            BeaconScreens.ARTICLE_SCREEN,
            arrayListOf(articleId)
        )
    }

    override fun logout(beaconId: String) {
        initBeacon(beaconId)
        Beacon.logout()
    }

    override fun reset(beaconId: String) {
        initBeacon(beaconId)
        Beacon.contactFormReset()
    }

    private fun initBeacon(beaconId: String) {
        Beacon.Builder()
            .withBeaconId(beaconId)
            .withLogsEnabled(true)
            .build()
    }

    private fun configureBeacon(beaconSettings: HelpScoutApiBeaconSettings) {
        initBeacon(beaconSettings.beaconId)
        Beacon.setConfigOverrides(
            BeaconConfigOverrides(
                docsEnabled = beaconSettings.docsEnabled,
                messagingEnabled = beaconSettings.messagingEnabled,
                chatEnabled = beaconSettings.messagingEnabled,
                color = beaconSettings.color?.hexCode,
                enablePreviousMessages = beaconSettings.enablePreviousMessages ?: true,
                focusMode = when (beaconSettings.focusModeOverride) {
                    null -> null
                    HelpScoutApiBeaconFocusMode.NEUTRAL -> FocusMode.NEUTRAL
                    HelpScoutApiBeaconFocusMode.ASK_FIRST -> FocusMode.ASK_FIRST
                    HelpScoutApiBeaconFocusMode.SELF_SERVICE -> FocusMode.SELF_SERVICE
                },
                contactForm = ContactFormConfigApi(
                    allowAttachments = beaconSettings.messagingSettings?.contactFormAllowAttachments
                        ?: true,
                    customFieldsEnabled = beaconSettings.messagingSettings?.contactFormCustomFieldsEnabled
                        ?: true,
                    showName = beaconSettings.messagingSettings?.contactFormShowNameField ?: true,
                    showSubject = beaconSettings.messagingSettings?.contactFormShowSubjectField
                        ?: true
                )
            ),
        )
    }
}